File: webmvp/src/lib/chordLayout.ts
/** Skyline layout for chord labels above lyrics (display-only; anchors unchanged in data). */

export type ChordLayoutPlacement = {
  left: number;
  tier: number;
};

export type ResolveChordLayoutInput = {
  /** Character start indices, sorted ascending. */
  starts: number[];
  /** Measured pixel anchor per chord (same order as starts). */
  anchorLeft: number[];
  /** Label width in px per chord (same order as starts). */
  width: number[];
  gap?: number;
  maxTiers?: number;
};

function overlaps(
  left: number,
  width: number,
  otherLeft: number,
  otherWidth: number,
  gap: number,
): boolean {
  return left < otherLeft + otherWidth + gap && left + width + gap > otherLeft;
}

function rightEdge(left: number, width: number): number {
  return left + width;
}

/**
 * Places chord labels on up to two tiers to avoid horizontal overlap.
 * Tier 0 stays at lyric anchors when possible; crowded chords move to tier 1.
 */
export function resolveChordLayout(
  input: ResolveChordLayoutInput,
): ChordLayoutPlacement[] {
  const {
    starts,
    anchorLeft,
    width,
    gap = 5,
    maxTiers = 2,
  } = input;

  const count = starts.length;
  if (count === 0) {
    return [];
  }

  const placed: ChordLayoutPlacement[] = [];
  const tierPlaced: ChordLayoutPlacement[][] = Array.from(
    { length: maxTiers },
    () => [],
  );

  for (let i = 0; i < count; i++) {
    const anchor = anchorLeft[i] ?? 0;
    const w = width[i] ?? 0;
    let assigned: ChordLayoutPlacement | null = null;

    for (let tier = 0; tier < maxTiers; tier++) {
      let left = anchor;

      if (tier === 1) {
        const onTier = tierPlaced[1];
        const lastOnTier = onTier[onTier.length - 1];
        if (lastOnTier) {
          const lastIdx = placed.indexOf(lastOnTier);
          const lastWidth = lastIdx >= 0 ? (width[lastIdx] ?? 0) : 0;
          left = Math.max(anchor, rightEdge(lastOnTier.left, lastWidth) + gap);
        }
      }

      const collides = tierPlaced[tier].some((p) => {
        const idx = placed.indexOf(p);
        const pw = idx >= 0 ? (width[idx] ?? 0) : 0;
        return overlaps(left, w, p.left, pw, gap);
      });

      if (!collides) {
        assigned = { left, tier };
        tierPlaced[tier].push(assigned);
        placed.push(assigned);
        break;
      }
    }

    if (!assigned) {
      const fallback = { left: anchor, tier: maxTiers - 1 };
      tierPlaced[maxTiers - 1].push(fallback);
      placed.push(fallback);
    }
  }

  return placed;
}

export function maxTierUsed(placements: ChordLayoutPlacement[]): number {
  if (placements.length === 0) {
    return 0;
  }
  return placements.reduce((max, p) => Math.max(max, p.tier), 0);
}

File: mobile/lib/domain/chord_layout.dart
class ChordLayoutPlacement {
  const ChordLayoutPlacement({required this.left, required this.tier});

  final double left;
  final int tier;
}

class ResolveChordLayoutInput {
  const ResolveChordLayoutInput({
    required this.starts,
    required this.anchorLeft,
    required this.width,
    this.gap = 5,
    this.maxTiers = 2,
  });

  final List<int> starts;
  final List<double> anchorLeft;
  final List<double> width;
  final double gap;
  final int maxTiers;
}

bool _overlaps(
  double left,
  double w,
  double otherLeft,
  double otherWidth,
  double gap,
) {
  return left < otherLeft + otherWidth + gap && left + w + gap > otherLeft;
}

double _rightEdge(double left, double width) => left + width;

List<ChordLayoutPlacement> resolveChordLayout(ResolveChordLayoutInput input) {
  final gap = input.gap;
  final maxTiers = input.maxTiers;
  final count = input.starts.length;
  if (count == 0) {
    return const [];
  }

  final placed = <ChordLayoutPlacement>[];
  final tierPlaced = List<List<ChordLayoutPlacement>>.generate(
    maxTiers,
    (_) => [],
  );

  for (var i = 0; i < count; i++) {
    final anchor = input.anchorLeft[i];
    final w = input.width[i];
    ChordLayoutPlacement? assigned;

    for (var tier = 0; tier < maxTiers; tier++) {
      var left = anchor;

      if (tier == 1) {
        final onTier = tierPlaced[1];
        if (onTier.isNotEmpty) {
          final lastOnTier = onTier.last;
          final lastIdx = placed.indexOf(lastOnTier);
          final lastWidth = lastIdx >= 0 ? input.width[lastIdx] : 0.0;
          left = left > _rightEdge(lastOnTier.left, lastWidth) + gap
              ? left
              : _rightEdge(lastOnTier.left, lastWidth) + gap;
        }
      }

      final collides = tierPlaced[tier].any((p) {
        final idx = placed.indexOf(p);
        final pw = idx >= 0 ? input.width[idx] : 0.0;
        return _overlaps(left, w, p.left, pw, gap);
      });

      if (!collides) {
        assigned = ChordLayoutPlacement(left: left, tier: tier);
        tierPlaced[tier].add(assigned);
        placed.add(assigned);
        break;
      }
    }

    if (assigned == null) {
      final fallback = ChordLayoutPlacement(left: anchor, tier: maxTiers - 1);
      tierPlaced[maxTiers - 1].add(fallback);
      placed.add(fallback);
    }
  }

  return placed;
}

int maxTierUsed(List<ChordLayoutPlacement> placements) {
  if (placements.isEmpty) {
    return 0;
  }
  return placements.fold(0, (max, p) => p.tier > max ? p.tier : max);
}
